Make MiqExpression#contains work in Rails 5.2

kbrock opened the pull request:
make MiqExpression#contains work in rails 5.2

MiqExpression CONTAINS built up the sql for a contains clause.
This would only work with direct simple relations and broke in the upgrade to rails 5.2.

In this PR, rails was asked to build the sql for a query that is close. Then the extra nodes are removed.
There were bind variables that needed to be removed. We typically use just compile the query and turn into a sql literal. Unfortunately, sql `IN` clauses are picky about the parameter that they accept, so we needed to pass arel.


The merge conditions code broke as well. We are patching it in this PR as well

kbrock commented:
^ fixed cops

jrafanie commented:
Will this replace this commit on the rails-5-2 branch?  https://github.com/jrafanie/manageiq/commit/1829461874def8692ae29b289ceb945e5f6d77f5

NickLaMuro commented:
Followup:  Is this Rails 5.2 only, or can we merge this into master now?

NickLaMuro commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
+    query_relation = main_model.joins(includes_associations).select(1).distinct
+    query = query_relation.arel
+
+    # 2. SELECT 1 FROM main_table [ ]
```
Should this comment be:

```ruby
# 2. SELECT 1 FROM [ ]  (drop main query)
```

Or, instead is the algorithm supposed to be:

```ruby
    # 2. drop joins from sub query
    # 3. replace the FROM with the secondary_table from the JOIN
```

Looks like we are doing the latter, meaning the comment above makes sense as is, but not sure.

kbrock replied:
agreed, this is confusing. I moved stuff around in attempt to address this

NickLaMuro commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -3415,4 +3415,59 @@
+    let(:eq_other_node) { Arel::Nodes::Equality.new GuestApplication.arel_table[:type], Arel::Nodes::SqlLiteral.new("abc") }
+    let(:eq_bind_node) { Arel::Nodes::Equality.new GuestApplication.arel_table[:type], Arel::Nodes::BindParam.new("") }
+
+    it "handles equals clauses" do
```
```ruby
.where("'vms'.id = 'guest_applications'.vm_or_template_id")
```

NickLaMuro commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -3415,4 +3415,59 @@
+      expect(query).to eq(nil)
+    end
+
+    it "handles equals bind clauses" do
```
```ruby
.where("'vms'.id = ?", "")
```

NickLaMuro commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -3415,4 +3415,59 @@
+      expect(query).to eq(nil)
+    end
+
+    it "handles reverse equals clauses" do
```
```ruby
# Note: second and third args are the same for .extract_key
.where("'guest_applications'.vm_or_template_id = 'vms'.id")
```

kbrock replied:
I was just putting stuff in them to see that it worked.

do you want me to actually build up different equal clauses?

NickLaMuro commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -3415,4 +3415,59 @@
+      expect(query).to eq(nil)
+    end
+
+    it "handles multi AND" do
```
```ruby
.where("'vms'.id = 'guest_applications'.vm_or_template_id AND 'guest_applications'.vm_or_template_id = 'abc' AND 'guest_applications'.vm_or_template_id = ''")
```

NickLaMuro commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
+    primary_attribute.in(query)
+  end
+
+  def extract_key(query, field1, field2)
```
I wrote a bunch of comments down in the specs, but I found myself having a hard time understanding what `.extract_key` was doing, and what it's return value was supposed to be.

From what I can tell, it is used to extract the "ON table_1.id = table_2.table_1_id" portion of the query, where ever it may be, but it is far from obvious that it is doing that.  Furthermore, it isn't obvious that `nil` means the query "only contains the `table_1.id = table_2.table_1_id` portion".

Some comments at a minimum would be nice, and adding more some comments in the specs to define what type of query is being test would also be welcome.

kbrock replied:
renamed this `remove_equality` and added comments

kbrock replied:
looking into converting back into a visitor

NickLaMuro commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
+    secondary_attribute = to_ref.klass.arel_attribute(join_keys.key)
+
+    # algorithm:
+    # 1. generate subquery as part of our main table since it has all the concepts we need.
```
Maybe note that `.select(1)` is added here.  It kind of gets lots in the noise in the code below, and then a `    query_relation.projections.shift` happens, and it isn't clear where it came from.

That, or maybe add a:

```ruby
query_relation.projections.shift  # drop placeholder `SELECT 1` (added in step 1)
```

NickLaMuro commented on lib/miq_expression.rb:
```diff
@@ -1399,13 +1403,7 @@ def to_arel(exp, tz)
         ids = tag.model.find_tagged_with(:any => parsed_value, :ns => tag.namespace).pluck(:id)
         tag.model.arel_attribute(:id).in(ids)
       else
-        raise unless field.associations.one?
```
Do we still need this?

kbrock replied:
we no longer limit it to relationships that are 1 deep.

NickLaMuro replied:
Does this make this branch a bit of an `enhancement` then?

kbrock replied:
thanks though - I added tests for this

kbrock replied:
also found a bug. (and added support for habtm and belongs_to)

kbrock replied:
I simplified it again (pushed off for another PR)

NickLaMuro commented:
@kbrock I think the PR description could really use some example cases that are being fixed here.  The specs you add only test the private methods, and don't really paint a picture at what the over "SQL" or `MiqExpression` might be.

kbrock commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
+    # 3. move the secondary_table from the JOIN to the FROM clause
+    # 4. add miq_expression filter to sub query.
+    # 5. if the join on clause has an AND, add the second part back to the query
+    # 6. replace SELECT 1 place holder with secondary_table id
```
thanks for this idea. I kinda ran with it and it became something I didn't expect

kbrock commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
-  def extract_where_values(klass, scope)
-    relation = ActiveRecord::Relation.new(klass, klass.arel_table, klass.predicate_builder)
-    relation = relation.instance_eval(&scope)
+  def subquery_for_contains(field, subquery_limiter)
```
done

kbrock commented on lib/miq_expression.rb:
```diff
@@ -1427,23 +1425,93 @@ def to_arel(exp, tz)
+    end
+  end
+
+  def traverse(query, &bind)
```
I couldn't get the visitor to generate a query tree. It seemed like it's goal was to generate a raw string. (I probably did it wrong)

kbrock commented:
heh. added a few more use cases and test
I'm having way too much fun with this
(I already have my next PR which adds multi level tagging.)

@NickLaMuro Yes this is an enhancement. By using rails to build the query for us, we now handle all sorts of use cases that we didn't before. I added tests to check the new use cases

If you need, I can rewrite the tests and commit before the main commit - so you can see the change.

NickLaMuro commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,74 @@
+      end
+    end
+
+    def self.traverse(query, &bind)
```
Potentially rename to `traverse_and_replace`?

kbrock replied:
will fix this. with intent to replace with a visitor

kbrock replied:
Made this change - left as the last commit. not sure if it made anything better.

kbrock commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,74 @@
+        else
+          query.class.new(children)
+        end
+      elsif query.kind_of?(Arel::Nodes::Unary)
```
I'm on the fence on this one. I changed this to Grouping and it is working for all our test cases.

kbrock commented:
Added the visitor.
fixed the remaining rubocop.

@NickLaMuro is this closer to what you had in mind?

NickLaMuro commented on spec/lib/miq_expression/query_helper_spec.rb:
```diff
@@ -0,0 +1,142 @@
+
+    # query:  "hosts"."type" = 'abc'
+    # result: "hosts"."type" = 'abc'
+    it "keeps a non matching query" do
```
"non matching query" reads oddly here, and doesn't really help clarify what `.remove_equality` does.

Maybe moving `vm_host_id.eq(host_id)` to a `let(:matching_query)` statement (since it is used in every example) would help clarify what "matching query" is supposed to mean.

NickLaMuro commented on spec/lib/miq_expression/query_helper_spec.rb:
```diff
@@ -0,0 +1,142 @@
+
+    # query:  "hosts"."type" = 'abc'
+    # result: "hosts"."type" = 'abc'
+    # This doesn't make sense, but we saw this example
```
"saw this example" where?  In specs?  In production?  In a magical fairy land where raw SQL is allowed to roam freely?

kbrock replied:
This is how the parameter was coming into our code for our code example

kbrock replied:
in miq test suite (thought it could have been another problem. who knows

NickLaMuro commented on spec/lib/miq_expression/query_helper_spec.rb:
```diff
@@ -0,0 +1,142 @@
+      expect(result).to be_nil
+    end
+
+    # query:  <invalid>
```
I feel like `<invalid>` is serving much of a purpose as to what `<invalid>` is supposed to be, or where it might happen.  Again, similar comment to the "but we saw this example" one.

NickLaMuro commented on spec/lib/miq_expression/query_helper_spec.rb:
```diff
@@ -0,0 +1,142 @@
+      end
+    end
+
+    # this helps us greate some of the odd cases we've seen around
```
"this helps us greate" ??

Is this supposed to be "generate"?  Probably not meant to be "great"...

kbrock replied:
create great case?

NickLaMuro commented on spec/lib/miq_expression_spec.rb:
```diff
@@ -498,23 +498,48 @@
     end
 
-    it "can't generate the SQL for a CONTAINS expression with [association.virtualassociation]" do
+    it "can't generate the SQL for a CONTAINS expression with invalid associations" do
```
Why the change here to something less specific?

kbrock replied:
in truth this case was passing an invalid association. had me totally confused working through it.

So I fixed this one and added what I think they meant to test a little below.

kbrock replied:
I moved it. hope the diff is more clear

kbrock commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,107 @@
+    # rubocop:disable Naming/MethodName (the visitor pattern is defined with odd naming conventions)
+    if ActiveRecord.version.to_s < "5.2"
+      # Rails 5.2 Arel::Visitors::Visitor
+      def accept(object, collector = nil)
```
just ensuring that `collector = nil` is here.
future versions all have the collector as a parameter

kbrock commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,107 @@
+class MiqExpression
+  class QueryHelper < Arel::Visitors::Visitor
+    # rubocop:disable Naming/MethodName (the visitor pattern is defined with odd naming conventions)
+    if ActiveRecord.version.to_s < "5.2"
```
I want to merge into master (5.1) and not rely upon having to be running 5.2

NickLaMuro replied:
Why?  Do you want to backport this?

Unless we find a reason to backport, let's just ~~git~~ get the MVP code necessary that is needed for 5.2 only instead of dragging a bunch of code along that will never get used in production anyway.

kbrock replied:
done - removed

kbrock commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,107 @@
+class MiqExpression
+  class QueryHelper < Arel::Visitors::Visitor
```
done

kbrock commented:
fixed rubocop directive.

made nick's suggested edits

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def visit_Arel_Nodes_Equality(object, collector)
+      field, binds = collector
```
Just to clarify:  `binds` here is getting mutated if `object.right` is a `Arel::Nodes::BindParam`, correct?  Same with `object` in that case as well?

Also, could we change `field` to `node_to_remove` here as well?  Would make it a bit more obvious what is being done here, and since `field` isn't part of the original method signature, I think that is fine (`object` on the other hand is and should probably stay as-is).

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+    # So converting a query into arel will loose the values for the bind parameters.
+    # The arel becomes incomplete and can not generate valid sql.
+    #
+    # This method takes the values from they query's bound_attributes and plugs them into the arel.
```
Typo:  "This method takes the values from **the~~y~~** query's ..."

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+    #
+    # Active record has the concept of the bind param values, but arel does not.
+    # So converting a query into arel will loose the values for the bind parameters.
+    # The arel becomes incomplete and can not generate valid sql.
```
It seems like the description of this method is describing the "why", not the "how".  Specifically, I think their are some concepts that are worth mentioning:

- `collector` is just a tuple of the `node_to_remove` and the `bound_attributes` that are passed in, and not a `ActiveRecord`/`Arel` construct.  I think this is important now more than when the `5.1` code existed because it is less obvious how the `collector` is being used now.
- Both `object` and `bind` will be mutated as a result of this method, not just returned.  Correct?

kbrock replied:
the first 2 lines tried to tell what was happening

kbrock replied:
changed collector to an actual class. still not collecting, but at least it is a single object.
in the future we may use the visitor to add node values to the collector

d-m-u replied:
wait can i also jump on the typo train here? sorry Keenan but it should be `lose`

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+         (object.right == field.left && object.left == field.right)
+        # remove this node from the query
+        nil
+      elsif object.right.kind_of?(Arel::Nodes::BindParam)
```
Should this be done in a separate `if` prior to the `if (object.left == ...` stuff above?  I am curious if there would ever be a case where the `Arel::Nodes::Equality` we are looking for would be behind a bind variable?  Is that possible or am I just being stupid?

(you are allowed to answer honestly to that last one)

kbrock replied:
For our use case, we are looking for a join between two tables. these should be 2 id columns, and I can't think of a reason why it would be a bind variable.

The extra scoping around this join is what we are keeping, which often does have a bind variable.

Removing an expression with a bind variable would cause issues. a bind should be popped in this case. I could add this case but it seems kinda out there.

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+    end
+
+    def visit_Arel_Nodes_And(object, collector)
```
Do we need to implement this ourselves?  Can't we just inherit from the `ToSql` visitor and avoid having to implement this method?

That aside, I am also not sure what this is doing that is special and needs to be implemented by us.  Is there something that I am missing?

NickLaMuro replied:
Seems like `Arel::Visitors::PostgreSQL` was used in the previous visitor (now deleted)

kbrock replied:
ToSql adds strings to the sql things like "AND" and OR
I suppose Descendant parser would probably be better

I could see how that turns out.

kbrock replied:
truth be told, I'm reluctant to use sql or even descent visitors. Nothing was built to allow someone to manipulate the tree. All of them are basically to produce sql.

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+    end
+
+    def visit_Arel_Nodes_Grouping(object, collector)
```
Same comment as was done with `visit_Arel_Nodes_And`.

kbrock commented on lib/miq_expression/query_helper.rb:
```diff
@@ -0,0 +1,99 @@
+
+      # Rails 5.2 Arel::Visitors::Visitor
+      # rubocop:disable Style/BlockDelimiters
+      def visit(object, collector)
```
turns out we were extending the wrong parent class.

Reducer is the proper one. It passes collector around

kbrock replied:
I also introduced a collector that is closer to their definition of collector. Hopefully it will be more clear

NickLaMuro commented on lib/miq_expression/remove_equality_visitor.rb:
```diff
@@ -0,0 +1,86 @@
+    end
+  end
+  # rubocop:disable Naming/MethodName
+  class RemoveEqualityVisitor < Arel::Visitors::Reduce
```
I don't think this exists in Rails 6... possibly in Rails 5.2 depending on what version of `ActiveRecord` we bumped to in the upgrade:

https://github.com/rails/arel/pull/512

NickLaMuro replied:
That said, it looks like it was effectively `Arel::Visitors::Visitor` anyway.

kbrock replied:
we can cross that bridge when we upgrade to 6.0

NickLaMuro replied:
This is a "I did a `bundle open activerecord` and couldn't find this class", not a "this might be an issue in the future" thing.  I see this being a problem right now, not later.

kbrock replied:
https://github.com/rails/rails/blob/5-2-stable/activerecord/activerecord.gemspec#L34

https://github.com/rails/arel/blob/v9.0.0/lib/arel/visitors/reduce.rb

5-1 uses v8 ==> https://github.com/rails/arel/blob/v8.0.0/lib/arel/visitors/reduce.rb

I think it is good for rails 5.1 and 5.2.
rails 6.0 just uses visitor

kbrock commented:
updated the typo and rebased to recent version of miq. hoping that fixes the rubygem fetch issue

kbrock commented:
sorry nick. this contains a few rails 5.2 vs 5.1 switches.
once we get this all set, I can remove them. I just wanted this off of master and the other one off of 5.2.

also, thinking I'll drop the traverser commit.  Not really doing any good besides making it difficult to rebase stuff.

NickLaMuro commented:
> also, thinking I'll drop the traverser commit. Not really doing any good besides making it difficult to rebase stuff.

That is fine.  What I tend to do, though (and you don't have to), is copy the branch prior to dropping a commit like that, and push it to my fork for historical reference, and link it in the PR (assuming you don't prune it later).  That way there is a way for someone to take a look at what was considered previously when looking back on this PR.

NickLaMuro commented:
> sorry nick. this contains a few rails 5.2 vs 5.1 switches.

Okay, if that is the case, I won't be reviewing further then.  I have made my case against this multiple times, and still disagree with you on this.  I don't think it will be productive having me review further.

kbrock commented:
reverted this PR back to previous form with a visitor.

I prefer new way #20110

NickLaMuro commented:
Closing this since this is the old form of the PR.
